ixnetwork/ixn_port.py

```diff
@@ -51,34 +51,47 @@ def reserve(self, location=None, force=False, wait_for_up=True, timeout=80):
             self.wait_for_up(timeout)
 
     def release(self):
-        if self.get_attribute('connectedTo') != self.api.null:
+        def _release_if_connected():
+            if self.get_attribute('connectedTo') == self.api.null:
+                return
             self.set_attributes(commit=True, connectedTo=self.api.null)
-        self.execute('releasePort', [self.ref])
-        self.wait_for_states(4, 'unassigned')
+            self.execute('releasePort', [self.ref])
+
+        self.wait_for_states(40, 'unassigned', wait_callback=_release_if_connected)
 
-    def wait_for_up(self, timeout=40):
+    def wait_for_up(self, timeout: Optional[int] = 40):
         """ Wait until port is up and running, including all parameters (admin state, oper state, license etc.).
 
         :param timeout: max time to wait for port up.
         """
         self.wait_for_states(timeout, 'up')
 
-    def wait_for_states(self, timeout: Optional[int] = 40, *states):
+    def wait_for_states(self, timeout: Optional[int] = 40, *states, wait_callback: Optional[Callable] = type(None)):
         """ Wait until port reaches one of the requested states.
 
         :param timeout: max time to wait for requested port states.
+        :param states: list of states to wait for.
+        :param wait_callback: callback to invoke at the start of each wait cycle.
         """
-
-        state = self.get_attribute('state')
-        for _ in range(timeout):
+        for n in range(1, timeout+1):
+            wait_callback()
```

Comment:
I think that using callbacks, while elegant, is somewhat overkill in this case and makes the code less readable.
I understand what you are trying to do, and actually, this is something that bugged me for a long time, so it seems right but I would go for a simpler solution using retries.
Something like:
def release(self):
    for _ in range(4):
        if self.get_attribute('connectedTo') != self.api.null:
            self.set_attributes(commit=True, connectedTo=self.api.null)
        self.execute('releasePort', [self.ref])
        try:
            self.wait_for_states(4, 'unassigned')
            return
        except TgnError as e:
             pass
    raise TgnError from e

What do you think?
Anyway, this is you change so I accept your call on this :)